kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -79,6 +83,126 @@ public static void validateSchema(StructType schema, boolean isColumnMappingEnab
     validateSupportedType(schema);
   }
 
+  /**
+   * Performs the following validations on an updated table schema using the current schema as a
+   * base for validation. ColumnMapping must be enabled to call this
+   *
+   * <p>The following checks are performed:
+   *
+   * <ul>
+   *   <li>No duplicate columns are allowed
+   *   <li>Column names contain only valid characters
+   *   <li>Data types are supported
+   *   <li>No new non-nullable fields are added or no tightening of nullable fields
+   *   <li>Physical column name consistency is preserved in the new schema
+   *   <li>Partition columns are not modified
+   *   <li>Nested IDs for array/map types are preserved in the new schema
+   *   <li>No type changes
+   * </ul>
+   */
+  public static void validateUpdatedSchema(
+      StructType currentSchema,
+      StructType newSchema,
+      Set<String> partitionCols,
+      Metadata metadata) {
+    checkArgument(
+        isColumnMappingModeEnabled(
+            ColumnMapping.getColumnMappingMode(metadata.getConfiguration())));
+    validateSchema(newSchema, true /*columnMappingEnabled*/);
+    Map<String, StructField> currentColumnPathToField = columnPathToField(currentSchema);
+    Map<String, StructField> updatedColumnPathToField = columnPathToField(newSchema);
+    validatePartitionColumns(newSchema, new ArrayList<>(partitionCols));
+    ColumnMapping.validateChangedColumnMapping(
+        currentColumnPathToField,
+        updatedColumnPathToField,
+        TableConfig.ICEBERG_COMPAT_V2_ENABLED.fromMetadata(metadata));
+    validateUpdatedSchemaCompatibility(currentColumnPathToField, updatedColumnPathToField);
+  }
+
+  /* Verifies that no non-nullable fields are added, no type changes are performed and that
+  partition columns are not modified */
+  private static void validateUpdatedSchemaCompatibility(
+      Map<String, StructField> currentColumnPathToStruct,
+      Map<String, StructField> updatedColumnPathToStruct) {
```

Comment:
It seems like we do all these checks using logical name paths instead of physical names or fieldIds? That seems wrong?... maybe I'm misinterpreting?

Reply:
This grouping is just for collecting the fields (nested and leaf). The actual validations use the field IDs/physical names

Reply:
Ah no, you are right. This was the case above, but this should use field IDs for the schema compatibility not the column paths

Reply:
See https://github.com/delta-io/delta/pull/4196#discussion_r1996197912 for test added to catch this case